### PR TITLE
Non-record: Does SLOT violate causal dependence? (empirical test + question)

### PR DESCRIPTION
**Pull request by andrewbaggio1**
## What this is

I wrote a script with a coding agent that tests whether SLOT violates causal
dependence. I think the answer is yes.

## My Test

- Take a window of tokens, run SLOT, and record the NLL at each scored position.
- Flip a single target token somewhere in the window and re-run SLOT.
- If the NLL changes at other positions, those predictions depended on the token
  we flipped.

Without SLOT, changing a target has no effect on other positions because the
model is causal. With SLOT, every single scored position is affected, and I
found a 100% violation rate across 240 tested pairs.

I also checked self-prediction: is P(x_{t+1}) better when x_{t+1} is actually
in the optimization targets vs when it's swapped for a random token? Yes, by
+0.24 nats (shared delta) and +0.73 nats (per-sample + logit bias).

I'm flagging my own submission (PR #1209).

I looked at PR #1229, which pushes SLOT to its logical extreme: per-sample
delta, per-sample logit bias, and scored-position masking. It gets 0.9300 BPB,
which is the best on the non-verified leaderboard and 0.19 below merged SOTA.
In defense of it, the mechanism is the same as shared-delta SLOT, just with
more parameters to memorize the evaluation targets.

## Counterargument

@AnubhavBharadwaaj correctly points out that in stride=64 sliding window,
1984/2048 tokens are already-scored context. So ~97% of the gradient comes
from known tokens. I think this is fair because the leakage in shared-delta
SLOT is small. But it's not zero, and "a little bit of future information"
is still future information.

## Reproducing

```bash
# No GPU needed. ~30 seconds on CPU.
python prove_slot_causal_violation.py
```

## Request for a decision

@0hq @valerio-oai SLOT has been debated across PRs #1084, #1128, #1172,
#1176, #1209 without a ruling. I'd really appreciate you weighing in!

Full writeup in the README generated by Claude Code.

### CODE REVIEW

**Comment by andrewbaggio1** [comment]
Closing — research documented, moving on to new approaches.

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Does SLOT violate causal dependence? (empirical test + question)

**BPB:** 0.9300 | **Compliance:** LOOKS CLEAN — score-first-per-chunk TTT (legal #1413 dexhunter pattern)

**What I found in the code** (head SHA `26136c0e79ff`, file `records/track_10min_16mb/2026-04-01_FullGPTQ_ScoreFirstTTT_SLOT_8xH100/train_gpt.py`):

The TTT path at line **1340** implements the score-first-per-chunk pattern: each chunk is scored under `torch.no_grad()` / `inference_mode()` before the `base_model.train()` + SGD adaptation runs on that same chunk, with an `is_last_chunk` guard so the final chunk gets no adaptation pass. This is the structural shape of the current leaderboard's legal frontier (PR #1413 dexhunter, the 1.0828 SP8192 + QK-Gain 5 + Legal TTT entry — verified at its head SHA against the `is_last_chunk` + `torch.no_grad()` score-first accumulator pattern).

Per **Issue #402 and Issue #677**, TTT is legal when each token is scored before the adapter updates on it, and that's what the code does here — chunk `ci` is scored under weights adapted only on chunks `0..ci-1`. No `prequant_ttt_adapt_adamw(val_tokens, ...)` multi-epoch fine-tune, no scored-region SLOT, no target-in-key n-gram cache.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.07s, dim=512, layers=11, vocab=1024, code=106952 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending standard checks (3-seed validation, 16MB artifact cap, 10-min wallclock on 8×H100 SXM). The compliance picture matches the legal reference frontier and no flags were raised by the classification pass.

*Auto-classification caveat:* this review was drafted by the AST-based classifier against a template derived from manually-reviewed cluster PRs (#1420, #1450, #1487, #1541, #1529, #1533, #1518). If I've misread a subtlety in your eval path — e.g., multi-epoch TTT that I mistook for single-pass, or a target-in-key lookup I missed in a helper function — please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.07s, dim=512, layers=11, vocab=1024, code=106952 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
